miku/src/main/scala/org/http4s/miku/NettyModelConversion.scala

```diff
@@ -138,34 +137,123 @@ object NettyModelConversion {
+  )(
+      implicit F: Effect[F],
+      ec: ExecutionContext
+  ): F[DefaultHttpResponse] =
+    if (httpRequest.headers.exists(
+          h => h.name.toString.equalsIgnoreCase("Upgrade") && h.value.equalsIgnoreCase("websocket")
+        )) {
+      val wsProtocol  = if (httpRequest.isSecure.exists(identity)) "wss" else "ws"
+      val wsUrl       = s"$wsProtocol://${httpRequest.serverAddr}${httpRequest.pathInfo}"
+      val bufferLimit = 65535 //Todo: Configurable. Probably param
+      val factory     = new WebSocketServerHandshakerFactory(wsUrl, "*", true, bufferLimit)
+      StreamSubscriber[F, WebSocketFrame].flatMap { subscriber =>
+        F.delay {
+          val processor = new Processor[WSFrame, WSFrame] {
+            def onError(t: Throwable): Unit = subscriber.onError(t)
+
+            def onComplete(): Unit = subscriber.onComplete()
+
+            def onNext(t: WSFrame): Unit = subscriber.onNext(nettyWsToHttp4s(t))
+
+            def onSubscribe(s: Subscription): Unit = subscriber.onSubscribe(s)
+
+            def subscribe(s: Subscriber[_ >: WSFrame]): Unit =
+              wsContext.webSocket.send.map(wsbitsToNetty).toUnicastPublisher().subscribe(s)
+          }
+
+          F.runAsync(Async.shift[F](ec) >> subscriber.stream.through(wsContext.webSocket.receive).compile.drain)(
+              _ => IO.unit
+            )
+            .unsafeRunSync()
+          val resp =
+            new DefaultWebSocketHttpResponse(httpVersion, HttpResponseStatus.OK, processor, factory)
+          wsContext.headers.foreach(h => resp.headers().add(h.name.toString(), h.value))
+          resp
+        }
+      }
+    } else {
+      wsContext.failureResponse.map(toNonWSResponse[F](_, httpVersion))
+    }
+
+  private def wsbitsToNetty(w: WebSocketFrame): WSFrame =
+    w match {
+      case Text(str, last)    => new TextWebSocketFrame(last, 0, str)
+      case Binary(data, last) => new BinaryWebSocketFrame(last, 0, Unpooled.wrappedBuffer(data))
+      case Ping(data)         => new PingWebSocketFrame(Unpooled.wrappedBuffer(data))
+      case Pong(data)         => new PongWebSocketFrame(Unpooled.wrappedBuffer(data))
+      case Continuation(data, last) =>
+        new ContinuationWebSocketFrame(last, 0, Unpooled.wrappedBuffer(data))
+      case Close(data) => new CloseWebSocketFrame(true, 0, Unpooled.wrappedBuffer(data))
+    }
+
+  private def nettyWsToHttp4s(w: WSFrame): WebSocketFrame =
+    w match {
+      case c: TextWebSocketFrame         => Text(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: BinaryWebSocketFrame       => Binary(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: PingWebSocketFrame         => Ping(bytebufToArray(c.content()))
+      case c: PongWebSocketFrame         => Pong(bytebufToArray(c.content()))
+      case c: ContinuationWebSocketFrame => Continuation(bytebufToArray(c.content()), c.isFinalFragment)
+      case c: CloseWebSocketFrame        => Close(bytebufToArray(c.content()))
+    }
 
```

Comment:
These casts are nasty

Reply:
in what way could we improve this though?